server/server.go

```diff
@@ -222,47 +222,46 @@ func (s *Server) handleGetMh(w lookupResponseWriter, r *http.Request) {
 	log := logger.With("multihash", mh)
 	ctx := context.Background()
 
+	// create result and error channels
 	resChan := make(chan model.ProviderResult)
-	errChan := make(chan error)
+	// the size of two is required so that the client doesn't in its go routine when trying to submit an error and return
+	errChan := make(chan error, 2)
 
+	// launch the find in a separate go routine
 	go s.c.FindAsync(ctx, mh, resChan, errChan)
 
 	haveResults := false
-	for {
-		select {
-		case err, ok := <-errChan:
-			if ok {
-				s.handleError(w, err, log)
-				return
-			}
-		case res, ok := <-resChan:
-			if !ok {
-				// If there were no results - return 404, otherwise finalize the response and return 200.
-				if !haveResults {
-					http.Error(w, "", http.StatusNotFound)
-					return
-				}
-				if err := w.Close(); err != nil {
-					switch e := err.(type) {
-					case errHttpResponse:
-						e.WriteTo(w)
-					default:
-						log.Errorw("Failed to finalize lookup results", "err", err)
-						http.Error(w, "", http.StatusInternalServerError)
-					}
-				}
-				return
-			}
-			// if this is the first result that we get - report latency as a time to first result
-			if !haveResults {
-				s.reportLatency(start, 200, r.Method, methodMultihash, true)
-			}
-			haveResults = true
-			if err := w.WriteProviderResult(res); err != nil {
-				log.Errorw("Failed to encode provider result", "err", err)
-				http.Error(w, "", http.StatusInternalServerError)
-				return
-			}
+
+	for pr := range resChan {
+		// if this is the first result that we got - report latency as a time to first result
+		if !haveResults {
+			s.reportLatency(start, 200, r.Method, methodMultihash, true)
+		}
+		haveResults = true
+		if err := w.WriteProviderResult(pr); err != nil {
+			log.Errorw("Failed to encode provider result", "err", err)
+			http.Error(w, "", http.StatusInternalServerError)
+			return
```

Comment:
Should an error be returned here, or shoud this error be logged and then continue processing results?  If there are other good results, it seems those should be returned.

Reply:
Well most likely the error would be because the client is closed already, so there is no point to continue (timeout)

Reply:
The problem is that returning here leaves `resChan` unread, and the goroutine is still writing to `resChan`. This means the goroutine will be blocked forever trying to write to `resChan`, and result in a resource leak.